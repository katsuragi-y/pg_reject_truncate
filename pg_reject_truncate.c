#include "postgres.h"

#include "tcop/utility.h"
#include "nodes/nodes.h"
#include "fmgr.h"

PG_MODULE_MAGIC;

static ProcessUtility_hook_type prev_ProcessUtility = NULL;

extern ProcessUtility_hook_type ProcessUtility_hook;

void _PG_init(void);
void _PG_fini(void);
static void pgrt_ProcessUtility(PlannedStmt *pstmt, const char *queryString,
                                ProcessUtilityContext context, ParamListInfo params,
                                QueryEnvironment *queryEnv,
                                DestReceiver *dest, QueryCompletion *qc);

void
_PG_init(void)
{
	prev_ProcessUtility = ProcessUtility_hook;
	ProcessUtility_hook = pgrt_ProcessUtility;
}

void
_PG_fini(void)
{
	ProcessUtility_hook = prev_ProcessUtility;
}

static void
pgrt_ProcessUtility(PlannedStmt *pstmt, const char *queryString,
                                ProcessUtilityContext context, ParamListInfo params,
                                QueryEnvironment *queryEnv,
                                DestReceiver *dest, QueryCompletion *qc)
{
	Node	*parsetree = pstmt->utilityStmt;

	/*
	 * if statements is truncate, rejecet it
	 */
	if (nodeTag(parsetree) == T_TruncateStmt)
	{
	    ereport(ERROR,
    	        (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
        	     errmsg("TRUNCATE IS NOT ALLOWED")));
	}

	standard_ProcessUtility(pstmt, queryString,
							context, params, queryEnv,
							dest, qc);
}
