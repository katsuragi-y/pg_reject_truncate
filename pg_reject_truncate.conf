shared_preload_libraries = 'pg_reject_truncate'
