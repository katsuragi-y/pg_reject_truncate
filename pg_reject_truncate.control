# pg_reject_truncate extension
default_version = '1.0'
module_pathname = '$libdir/pg_reject_truncate'

