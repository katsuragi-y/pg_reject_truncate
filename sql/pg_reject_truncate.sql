CREATE EXTENSION pg_reject_truncate;
LOAD 'pg_reject_truncate';

---
--- set up a table
---
CREATE TABLE pg_reject_truncate_test (
	id INTEGER
);
INSERT INTO pg_reject_truncate_test VALUES(1);

--- check the contents of the table
SELECT * FROM pg_reject_truncate_test;

--- truncate
TRUNCATE pg_reject_truncate_test;

--- check table still exists and  recheck the contents of the table
SELECT * FROM pg_reject_truncate_test;

DROP EXTENSION pg_reject_truncate;
