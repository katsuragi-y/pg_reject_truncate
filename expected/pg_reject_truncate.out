CREATE EXTENSION pg_reject_truncate;
---
--- set up a table
---
CREATE TABLE pg_reject_truncate_test (
	id INTEGER
);
INSERT INTO pg_reject_truncate_test VALUES(1);
--- check the contents of the table
SELECT * FROM pg_reject_truncate_test;
 id 
----
  1
(1 row)

--- truncate
TRUNCATE pg_reject_truncate_test;
ERROR:  TRUNCATE IS NOT ALLOWED
--- check table still exists and  recheck the contents of the table
SELECT * FROM pg_reject_truncate_test;
 id 
----
  1
(1 row)

DROP EXTENSION pg_reject_truncate;
